[Elroy Goh] IP

elroygohjy opened the pull request:
### Duke
> "Without continual growth and progress, such words as improvement, achievement, and success have no meaning."                
**Benjamin Franklin**

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/elroygohjy/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

- [ ]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    } 
}
```

elroygohjy committed:
add Level-1

elroygohjy committed:
Add Level-2

elroygohjy committed:
Add Level-3 and A-Classes

elroygohjy committed:
Add Level-4 and A-Inheritance

elroygohjy committed:
Add Level-4 and A-Inheritance

elroygohjy committed:
Add A-TextUiTesting

elroygohjy committed:
add Level-5 and A-Exceptions

elroygohjy committed:
Add Level-6 and A-Collections

elroygohjy committed:
Change the filename from RobotFriend.java to Duke.java

elroygohjy committed:
Add Level-7

elroygohjy committed:
Add Level-8

elroygohjy committed:
Merge branch 'branch-Level-7'

elroygohjy committed:
Merge branch 'branch-Level-8'

elroygohjy committed:
Add A-Packages and A-MoreOOP

elroygohjy committed:
Add A-JUnit

elroygohjy committed:
Refactor package names

elroygohjy committed:
A-JavaDoc

elroygohjy committed:
Add A-CodingStandard

elroygohjy committed:
Add A-Level-9

elroygohjy committed:
Merge branch 'branch-A-JavaDoc'

elroygohjy committed:
Merge branch 'branch-A-CodingStandard'

elroygohjy committed:
Merge branch 'branch-Level-9'

elroygohjy committed:
Add A-Gradle

elroygohjy committed:
Merge branch 'branch-A-Gradle'

elroygohjy committed:
Add A-CheckStyle and fix all checkstyle issues

elroygohjy committed:
Merge branch 'branch-A-CheckStyle'

lizchow commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+package duke;
```
I dont think file naming of packages should be capitalised? So it should be `src/main/java/duke/Duke.java` instead of `src/main/java/Duke/Duke.java`. Same for the `Exception` packages, should be `exception` instead.

lizchow commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+			default:
+				throw new InvalidCommandException();
+			}
+		} catch (InvalidCommandException | NoTaskDescriptionException | NoDateIndicatorException
+				| NoDateException | InvalidFindException err) {
```
Instead of catching so many exceptions, I think it might be better to let all these exceptions extend from `DukeException` and just catch `DukeException`?

lizchow commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+package duke;
+
+import duke.exception.*;
```
I think wildcard imports are not allowed according to the module coding standards?

lizchow commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	/**
+	 * Return true if parser read user input of "bye".
```
```suggestion
	 * Returns true if parser read user input of "bye".
```

lizchow commented on src/main/java/Duke/Task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+	 * The deadline in day.
+	 */
+	protected LocalDate by;
+	private static final DateTimeFormatter inputDateFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy");
```
I believe this is a constant, so it would be more appropriate to name `INPUT_DATE_FORMAT`? Same as the `outputDateFormat` below.

lizchow commented on src/main/java/Duke/Task/Event.java:
```diff
@@ -0,0 +1,54 @@
+		return "E" + " | " + completeBinary + " | " + this.taskDetails + " | " + this.at.format(inputDateFormat);
+	}
+
+}
```
This file seems to be missing a newline at end of file. I think there are a few more files with missing newline at end of file.
```suggestion
}

```

lizchow commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,235 @@
+			} catch (NumberFormatException err) {
+				throw new InvalidIndexException("done");
+			}
+			if (!isTaskDoneCommand || taskNumber > tasks.size()) {
+				throw new InvalidIndexException("done");
+			}
+		} catch (InvalidIndexException err) {
+			UI.printErrorMessage(err);
+			return;
+		}
```
Generally exception as control flow (i.e. throwing an exception above and catching it a few lines below) is not a good idea. Maybe a better way for this try-catch block might be:  
```
try {
    taskNumber = Integer.parseInt(tokens[1]);
    if (!isTaskDoneCommand || taskNumber > tasks.size()) {
        throw new InvalidIndexException("done");
    }
catch (NumberFormatException err) {
     throw new InvalidIndexException("done");
}
```
and catch `InvalidIndexException` in `src/main/java/Duke/Parser.java`.

lizchow commented on src/main/java/Duke/UI.java:
```diff
@@ -0,0 +1,103 @@
+    public static void bye() {
+        System.out.println(LINE);
+        System.out.println(ROBOT_ICON + ": " + "Bye~ Hope to see you soon :)");
+        System.out.println(LINE);
```
Not sure if I've missed it, but it might be good to close the scanner here.

elroygohjy committed:
Add Level-10

yzhilong commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+                    throw new DukeException("Invalid command! Try again.");
+                }
+                return tasks.listToString();
+
```
```suggestion
```
I think this new line is unnecessary? Since it is within the same logical block of parsing the token

yzhilong replied:
```suggestion
```
I think this new line is unnecessary? Since it is within the same logical block of parsing the token

yzhilong commented on src/main/java/Duke/Task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * The deadline in day.
+     */
+    protected LocalDate by;
```
```suggestion
    protected LocalDate deadlineDate;
```
I was a little confused by what `by` referred to until I read the rest of your code. Maybe calling it `deadlineDate` would make the purpose of this variable clearer?

yzhilong replied:
```suggestion
    protected LocalDate deadlineDate;
```
I was a little confused by what `by` referred to until I read the rest of your code. Maybe calling it `deadlineDate` would make the purpose of this variable clearer?

yzhilong commented on src/main/java/Duke/UI.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * UI for Duke.
+ */
+public class UI {
```
```suggestion
public class Ui {
```
I think abbreviations are not supposed to be uppercase when they are part of a name?

alinaleehx commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+    public Parser(TaskList tasks) {
+        this.tasks = tasks;
+        this.isExit = false;
+    }
```
I think you need a new line here.

alinaleehx commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+        tasks.addTask(newTask);
+        return "You have added this following task to the list: \n" + newTask.toString() + "\n"
+                + "You have " + tasks.noOfTasks() + " tasks now.";
+    }
```
A new line between line 66 and 67 would be good too.

alinaleehx commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,98 @@
+
+
+
+
```
You could add javadoc (header comments) for public classes like Storage.

alinaleehx commented on src/main/java/gui/Main.java:
```diff
@@ -0,0 +1,31 @@
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/views/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
```
Maybe you can use a more descriptive name instead of ap.

alinaleehx commented on src/main/java/Duke/Task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Deadline-type Task is a task consisting of its details and deadline in day.
+ */
```
Suggested change:
```suggestion
/**
 * The Deadline class extends the Task class and encapsulates a deadline task 
 *  consisting of details and deadline in a day.
 */
```

alinaleehx reviewed (commented):
A great job overall🥇 👍 Just a few nits to fix!

elroygohjy committed:
Add A-Assertions

Assertion will make sure that there will be no unexpected behaviour to tasks when there is invalid command

elroygohjy committed:
Add A-CodeQuality

elroygohjy committed:
Merge branch 'branch-A-CodeQuality'

elroygohjy committed:
Add C-MassOps

elroygohjy committed:
Merge pull request #1 from elroygohjy/branch-A-Assertions

Add A-Assertions

elroygohjy committed:
Merge pull request #2 from elroygohjy/branch-A-CodeQuality

Add A-CodeQuality

elroygohjy committed:
Merge tag 'C-MassOps'

elroygohjy committed:
fix checkstyle, improve Gui and add README

elroygohjy committed:
fix README

elroygohjy committed:
fix README exit feature and add UI.png

elroygohjy committed:
change font and refactor package name

elroygohjy committed:
fix error where error for invalid date time is not catch
